relay-server/src/services/server.rs

```diff
@@ -55,115 +40,204 @@ const CLIENT_HEADER_TIMEOUT: Duration = Duration::from_secs(5);
-        // Inform the user about a removed feature.
-        if config.tls_listen_addr().is_some()
-            || config.tls_identity_password().is_some()
-            || config.tls_identity_path().is_some()
-        {
-            return Err(ServerError::TlsNotSupported);
-        }
+/// Indicates the type of failure of the server.
+#[allow(clippy::enum_variant_names)]
+#[derive(Debug, thiserror::Error)]
+pub enum ServerError {
+    /// Binding failed.
+    #[error("bind to interface failed")]
+    BindFailed(#[from] io::Error),
+
+    /// TLS support was not compiled in.
+    #[error("SSL is no longer supported by Relay, please use a proxy in front")]
+    TlsNotSupported,
+}
 
-        Ok(Self { config, service })
+type App = NormalizePath<axum::Router>;
+
+/// Build the axum application with all routes and middleware.
+fn make_app(service: ServiceState) -> App {
+    // Build the router middleware into a single service which runs _after_ routing. Service
+    // builder order defines layers added first will be called first. This means:
+    //  - Requests go from top to bottom
+    //  - Responses go from bottom to top
+    let middleware = ServiceBuilder::new()
+        .layer(axum::middleware::from_fn(middlewares::metrics))
+        .layer(CatchPanicLayer::custom(middlewares::handle_panic))
+        .layer(SetResponseHeaderLayer::overriding(
+            header::SERVER,
+            HeaderValue::from_static(constants::SERVER),
+        ))
+        .layer(SetResponseHeaderLayer::overriding(
+            HeaderName::from_static("cross-origin-resource-policy"),
+            HeaderValue::from_static("cross-origin"),
+        ))
+        .layer(NewSentryLayer::new_from_top())
+        .layer(SentryHttpLayer::with_transaction())
+        .layer(middlewares::trace_http_layer())
+        .map_request(middlewares::remove_empty_encoding)
+        .layer(RequestDecompressionLayer::new())
+        .layer(
+            CompressionLayer::new()
+                .compress_when(SizeAbove::new(COMPRESSION_MIN_SIZE).and(DefaultPredicate::new())),
+        );
+
+    let router = crate::endpoints::routes(service.config())
+        .layer(middleware)
+        .with_state(service);
+
+    // Add middlewares that need to run _before_ routing, which need to wrap the router. This are
+    // especially middlewares that modify the request path for the router:
+    NormalizePath::new(router)
+}
+
+fn listen(config: &Config) -> Result<TcpListener, ServerError> {
+    // Inform the user about a removed feature.
+    if config.tls_listen_addr().is_some()
+        || config.tls_identity_password().is_some()
+        || config.tls_identity_path().is_some()
+    {
+        return Err(ServerError::TlsNotSupported);
     }
+
+    let addr = config.listen_addr();
+    let socket = match addr {
+        SocketAddr::V4(_) => TcpSocket::new_v4(),
+        SocketAddr::V6(_) => TcpSocket::new_v6(),
+    }?;
+
+    socket.bind(addr)?;
```

Comment:
I'm not setting keepalive on this socket. IIUC that's not needed as we set keepalive this is done later as we accept connections.

Reply:
Duplicate of https://github.com/getsentry/relay/pull/3959#discussion_r1734669755